### PR TITLE
feat(core): add function for verifying ethereum addresses in bitgo express

### PR DESCRIPTION
**Pull request by gianchandania**
**Background**

BitGo Express does not adequately verify the addresses returned by BitGo Servers while building transactions. If the BitGo servers are compromised, the address verification logic ensures that BitGo Express cannot be tricked by providing a wrong address. 

**Changes**

- Added function `verifyAddress()` for Ethereum.

- Added `calculateEthAddress2()`: function to create ethereum address for forwarder version 1, `calculateEthAddress()`: function to create ethereum address for forwarder version 0, `getProxyInitcode()`: function to get the initcode which is used to create address with forwarder version 1  in utils file

- Added constants for forwarderFactoryAddress and forwarderImplementationAddress in config file

- Added unit tests for the Address verification function

- Added assert package for test cases of address verification function

- Upgraded version of ethereumjs-util because generateAddress2 function which is used to calculate ethereum address for forwarder version 1 was not in the 4.x.x version. Upgraded the version to 6.2.1 to make it in sync with bitgo-microservices

- Make a minor change in the `getSendMethodArgs()` method in ERC20Token and Eth class because of ethereumjs-util's version update

Ticket: BG-33701

### CODE REVIEW

File: modules/core/src/v2/coins/eth.ts [hunk 1]
```diff
@@ -1617,7 +1636,63 @@ export class Eth extends BaseCoin {
...
+    let expectedAddress;
+    let actualAddress;
+
+    const { address, index, chain, coinSpecific, baseAddress } = params;
+
+    if (address && !self.isValidAddress(address)) {
+      throw new InvalidAddressError(`invalid address: ${address}`);
+    }
+
+    if ((_.isUndefined(chain) && _.isUndefined(index)) || !(_.isFinite(chain) && _.isFinite(index))) {
```
[the remark this answers is not in the document]
Comment (barathcj) [reply modules/core/src/v2/coins/eth.ts:1674]: agree. Lets remove this check. Index and chain are barely used (if at all) for eth

File: modules/core/src/v2/coins/eth.ts [hunk 2]
```diff
@@ -1617,7 +1636,63 @@ export class Eth extends BaseCoin {
...
+   * @param {Number} params.chain Derivation chain
+   * @param {Object} params.coinSpecific Coin-specific details for the address such as a forwarderVersion
+   * @param {String} params.baseAddress The base address of the wallet on the network
+   * @throws {InvalidAddressError}
+   * @throws {InvalidAddressDerivationPropertyError}
+   * @throws {InvalidAddressVerificationObjectPropertyError}
+   * @throws {UnexpectedAddressError}
+   * @returns {Boolean} True if address is valid
+   */
+  verifyAddress(params: VerifyEthAddressOptions): boolean {
```
[the remark this answers is not in the document]
Comment (barathcj) [reply modules/core/src/v2/coins/eth.ts:1658]: ah ok 👍

File: modules/core/src/v2/coins/erc20Token.ts [hunk 3]
```diff
@@ -326,7 +326,7 @@ export class Erc20Token extends Eth {
       {
         name: 'signature',
         type: 'bytes',
-        value: optionalDeps.ethUtil.toBuffer(txInfo.signature),
+        value: optionalDeps.ethUtil.toBuffer(optionalDeps.ethUtil.addHexPrefix(txInfo.signature)),
```
[the remark this answers is not in the document]
Comment (barathcj) [reply modules/core/src/v2/coins/erc20Token.ts:330]: ok

File: modules/core/src/v2/coins/eth.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gianchandania) [reply modules/core/src/v2/coins/eth.ts]: Done

File: yarn.lock [hunk 4]
```diff
@@ -3896,7 +3896,7 @@ bn.js@^3.1.1:
   resolved "https://registry.yarnpkg.com/bn.js/-/bn.js-3.3.0.tgz#1138e577889fdc97bbdab51844f2190dfc0ae3d7"
   integrity sha1-ETjld4if3Je72rUYRPIZDfwK49c=
 
-bn.js@^4.0.0, bn.js@^4.1.0, bn.js@^4.11.0, bn.js@^4.11.1, bn.js@^4.11.6, bn.js@^4.11.8, bn.js@^4.11.9, bn.js@^4.12.0, bn.js@^4.4.0, bn.js@^4.8.0:
+bn.js@^4.0.0, bn.js@^4.1.0, bn.js@^4.11.0, bn.js@^4.11.1, bn.js@^4.11.6, bn.js@^4.11.8, bn.js@^4.11.9, bn.js@^4.12.0, bn.js@^4.4.0:
```
Comment (twtaylorbitgo) [inline yarn.lock:3899]: why did we remove a BN dep here?

File: modules/core/src/v2/internal/util.ts [hunk 5]
```diff
@@ -197,4 +197,45 @@ export class Util {
...
+   * Take the implementation address for the proxy contract, and get the binary initcode for the associated proxy
+   * @param implementationAddress The address of the implementation contract for the proxy
+   * @return {string} Binary hex string of the proxy
+   */
+  static getProxyInitcode = (implementationAddress) => {
+    const target = ethUtil.stripHexPrefix(implementationAddress.toLowerCase()).padStart(40, '0');
+
+    // bytecode of the proxy, from:
+    // https://github.com/BitGo/eth-multisig-v4/blob/d546a937f90f93e83b3423a5bf933d1d77c677c3/contracts/CloneFactory.sol#L42-L56
+    return `0x3d602d80600a3d3981f3363d3d373d3d3d363d73${target}5af43d82803e903d91602b57fd5bf3`;
```
Comment (twtaylorbitgo) [inline modules/core/src/v2/internal/util.ts:211]: the target here should be using rlp encoding instead of what we're doing here

**Review by twtaylorbitgo — CHANGES_REQUESTED** [review]

File: modules/core/src/v2/internal/util.ts (hunk 5 above)
Comment (gianchandania) [reply modules/core/src/v2/internal/util.ts:211]: I have taken this function from the microservices only, please let me know if it still needs to be changed. https://github.com/BitGo/bitgo-microservices/blob/develop/packages/wallet-platform/app/statics/ethContracts.js#:~:text=*/-,exports.getProxyInitcode%20%3D%20(implementationAddress)%20%3D%3E%20%7B,-const%20target%20%3D%20ethUtil

File: yarn.lock (hunk 4 above)
Comment (gianchandania) [reply yarn.lock:3899]: This got removed because ethereumjs-utils upgraded it's version of bn.js from 4.8.0 to 4.11.0

**Review by twtaylorbitgo — DISMISSED** [review]

**Review by twtaylorbitgo — APPROVED** [review]

**Review by twtaylorbitgo — CHANGES_REQUESTED** [review]
This actually needs a corresponding test in the bitgo clients... sorry. If you look at bitgo-client-private we are using this all over the sendPrepareDirective.

**Comment by barathcj** [comment]
> This actually needs a corresponding test in the bitgo clients... sorry. If you look at bitgo-client-private we are using this all over the sendPrepareDirective.

@twtaylorbitgo https://github.com/BitGo/bitgo-microservices/blob/66e9487d88ac90e848c5bcb4c191c7bd399b357e/packages/client-wallet-platform/bitgo/modules/v2/wallet/send/sendPrepare/walletSendPreparetxV2Directive.js#L606 Is this what you referring to? This seems to be for sending transactions and doing validity checks on the recipient field of a send. This PR is for verifying addresses that are generated by BitGo. These seem to be 2 different things (correct me if I'm wrong)
Regardless the client wallet-platform-tests are very frontend specific and I don't believe verifyAddress is really checked from there. Akash also not really seen it, do you think we should get some front end folks to look here?

**Review by twtaylorbitgo — APPROVED** [review]

**Pull request merged**